flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/SourceOperator.java

```diff
@@ -276,20 +277,33 @@ public void close() throws Exception {
     }
 
     @Override
-    public InputStatus emitNext(DataOutput<OUT> output) throws Exception {
+    public DataInputStatus emitNext(DataOutput<OUT> output) throws Exception {
         // guarding an assumptions we currently make due to the fact that certain classes
         // assume a constant output
         assert lastInvokedOutput == output || lastInvokedOutput == null;
 
         // short circuit the common case (every invocation except the first)
         if (currentMainOutput != null) {
-            return sourceReader.pollNext(currentMainOutput);
+            return convertToInternalStatus(sourceReader.pollNext(currentMainOutput));
         }
 
         // this creates a batch or streaming output based on the runtime mode
         currentMainOutput = eventTimeLogic.createMainOutput(output);
         lastInvokedOutput = output;
-        return sourceReader.pollNext(currentMainOutput);
+        return convertToInternalStatus(sourceReader.pollNext(currentMainOutput));
+    }
+
+    private DataInputStatus convertToInternalStatus(InputStatus inputStatus) {
+        switch (inputStatus) {
+            case MORE_AVAILABLE:
+                return DataInputStatus.MORE_AVAILABLE;
+            case NOTHING_AVAILABLE:
+                return DataInputStatus.NOTHING_AVAILABLE;
+            case END_OF_INPUT:
+                return DataInputStatus.END_OF_INPUT;
+            default:
+                throw new IllegalArgumentException("Unknown input status: " + inputStatus);
+        }
```

Comment:
I would be curious how well JIT can optimise this... probably as long as `sourceReader.pollNext()` is de-virtualised and inlined this will be for free. But otherwise I would guess it will cost a couple of CPU cycles. Either way I don't expect this to be visible as a performance regression, but I'm just curious...